Feature/improve logging of failures when in parallel mode

JonathanWylie opened the pull request:
Please give me a review before I make a pull request to upstream.

JonathanWylie committed:
When in parallel mode, log the errors for all the processes that failed, not just the first one.

JonathanWylie committed:
pep8 fixes

JonathanWylie committed:
test for logging behaviour

ScoreUnder commented on fabric/tasks.py:
```diff
@@ -411,14 +417,23 @@ def execute(task, *args, **kwargs):
                     else:
-                        error(err)
+                        failed = True
+                        error(name + "experienced an unknown error", func=failed_instance)
```
May need a space before `"experienced"`

ScoreUnder commented on fabric/tasks.py:
```diff
@@ -411,14 +417,23 @@ def execute(task, *args, **kwargs):
+                        failed = True
+                        error(name + "experienced an unknown error", func=failed_instance)
                 results[name] = d['results']
+            if failed:
```
This will suppress the warnings that should be given by the `NetworkError` check above.

I'd consider something like this, ±pep8:

``` python
            if failedMessages:
                error_func = None if failed else warn
                error(err + "\n\n" + "\n".join(failedMessages) + "\n\n............................", func=error_func)
```

JonathanWylie committed:
Report as warnings NetworkErrors if they are supposed to be suppressed.
 If there are other more serious errors from other hosts, then we report
 all the failures together, but possibly as an abort.

JonathanWylie committed:
Test for warn suppression of NetworkErrors
